### PR TITLE
Use a role cache to avoid separate locking paths

### PR DESCRIPTION
**Pull request by jefferai**
Due to the various locked/nonlocked paths we had a case where we weren't
always checking for secondary status before trying to upgrade. This
broadly simplifies things by using a cache to store the current role
values (avoiding a lot of storage hits) and updating the cache on any
write, delete, or invalidation.

### CODE REVIEW

File: builtin/credential/aws/path_role.go [hunk 1]
```diff
@@ -218,82 +220,81 @@ func pathListRoles(b *backend) *framework.Path {
...
+		return nil, errwrap.Wrapf("error upgrading roleEntry: {{err}}", err)
+	}
+	if needUpgrade && (b.System().LocalMount() || !b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary|consts.ReplicationPerformanceStandby)) {
+		if err = b.setRole(ctx, s, roleName, &result); err != nil {
+			return nil, errwrap.Wrapf("error saving upgraded roleEntry: {{err}}", err)
+		}
+	}
 
-	return b.nonLockedSetAWSRole(ctx, s, roleName, roleEntry)
+	return &result, nil
```
Comment (vishalnayak) [inline builtin/credential/aws/path_role.go]: Do we not want to insert the role in cache after reading from storage?

File: builtin/credential/aws/path_role.go [hunk 2]
```diff
@@ -512,16 +496,14 @@ func (b *backend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request
 			Version: currentRoleStorageVersion,
 		}
 	} else {
-		needUpdate, err := b.upgradeRoleEntry(ctx, req.Storage, roleEntry)
+		// We want to always use a copy so we arne't modifying items in the
```
Comment (vishalnayak) [inline builtin/credential/aws/path_role.go]:
```suggestion
		// We want to always use a copy so we aren't modifying items in the
```

File: builtin/credential/aws/path_role.go (hunk 1 above)
Comment (jefferai) [reply builtin/credential/aws/path_role.go]: Ah yes! Will fix.

**Comment by jefferai** [comment]
Yeah that seems unrelated.

**Comment by catsby** [comment]
> TestBackendAcc_LoginWithCallerIdentity

This test passes for me:

```
[vault][aws-role-cache](4)$ make testacc TEST=./builtin/credential/aws TESTARGS="-run=TestBackendAcc_LoginWithCallerIdentity"
==> Checking that build is using go version >= 1.11...
==> Using go version 1.12.6...
VAULT_ACC=1 go test -tags='' ./builtin/credential/aws -v -run=TestBackendAcc_LoginWithCallerIdentity -timeout=45m
=== RUN   TestBackendAcc_LoginWithCallerIdentity
--- PASS: TestBackendAcc_LoginWithCallerIdentity (3.03s)
PASS
ok      github.com/hashicorp/vault/builtin/credential/aws       3.262s
```

File: builtin/credential/aws/backend.go [hunk 3]
```diff
@@ -201,13 +206,16 @@ func (b *backend) periodicFunc(ctx context.Context, req *logical.Request) error
...
+	switch {
+	case key == "config/client":
 		b.configMutex.Lock()
 		defer b.configMutex.Unlock()
 		b.flushCachedEC2Clients()
 		b.flushCachedIAMClients()
 		b.defaultAWSAccountID = ""
+	case strings.HasPrefix(key, "role"):
+		// TODO: We could make this better
+		b.roleCache.Flush()
```
Comment (chrishoffman) [inline builtin/credential/aws/backend.go]: I appreciate the simplicity here :)

Comment (jefferai) [reply builtin/credential/aws/backend.go]: Yeah...I think we could just separate the role name and use that, but, this isn't far worse.

**Pull request merged**